vhost_rs/src/vhost_user/sock_ctrl_msg.rs

```diff
@@ -176,10 +176,13 @@ fn raw_recvmsg(fd: RawFd, iovecs: &mut [iovec], in_fds: &mut [RawFd]) -> Result<
 
     // Safe because the msghdr was properly constructed from valid (or null) pointers of the
     // indicated length and we check the return value.
-    let total_read = unsafe { recvmsg(fd, &mut msg, libc::MSG_WAITALL) };
+    let total_read = unsafe { recvmsg(fd, &mut msg, 0) };
 
     if total_read == -1 {
-        return Err(Error::last());
+        let total_read = unsafe { recvmsg(fd, &mut msg, libc::MSG_WAITALL) };
```

Comment:
This is not an acceptable fix IMO. With this patch, virtio-fs will still fail randomly when not all data will be read at once. The reason is that `total_read` won't be `-1`, but it won't be the full message either.
I think we're trying to solve the problem the wrong way here. Using `MSG_WAITALL` should always work since it's waiting for the full message to be received.
Which vhost-user-blk backend are you using? It seems to me like something might be wrong from the backend, or that we are missing something else from the `vhost_rs` crate.

Reply:
New patch to fix this, https://github.com/intel/cloud-hypervisor/pull/173/commits/009394c8e85200430e37fa10e3f9c06efa51c355